crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+                )
+                    .run_if(in_state(GameState::Playing)),
+            );
+
+        #[cfg(feature = "energy_graph_debug_lines")]
+        app.add_plugins(DebugLinesPlugin::default()).add_systems(
+            PostUpdate,
+            (debug_lines).run_if(in_state(GameState::Playing)),
+        );
+    }
+}
+
+/// wrapped entity to allow for default values (se we can work with TinyVec)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct NearbyEntity(Entity);
+
+impl Default for NearbyEntity {
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+impl From<NearbyEntity> for Entity {
+    fn from(val: NearbyEntity) -> Self {
+        val.0
+    }
+}
+
+impl From<Entity> for NearbyEntity {
+    fn from(entity: Entity) -> Self {
+        Self(entity)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone, Default)]
+pub struct Nearby {
+    receivers: TinyVec<[NearbyEntity; 32]>,
+    producers: TinyVec<[NearbyEntity; 32]>,
+}
+
+impl Nearby {
+    fn both(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.receivers
+            .clone()
+            .into_iter()
+            .chain(self.producers.clone())
+            .collect()
+    }
+
+    fn clear(&mut self) {
+        self.receivers.clear();
+        self.producers.clear();
+    }
+
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        self.receivers.retain(|&e| e != entity);
+        self.producers.retain(|&e| e != entity);
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: Nearby,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    pub fn units(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.units.both()
+    }
+}
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
```

Comment:
IMHO semantically proper way would be to trigger this on `Battery` rather than `Active`. Id est, anything with a battery can receive energy to charge it.